File: client/src/app/converters/bart_config.converter.ts
import { BartConfig } from "@app/models/bart_config";
import { IBartConfig } from "@app/proto/antigravity";

export class BartConfigConverter {
  static fromProto(proto: IBartConfig): BartConfig {
    return {
      name: proto.name || "BART",
      deviceName: proto.deviceName || "",
      deviceAddress: proto.deviceAddress || "",
      numLanes: proto.numLanes || 8,
      minLapMs: proto.minLapMs != null ? proto.minLapMs : 1,
      lapPinPitBehavior: proto.lapPinPitBehavior ?? 3,
      lapPinBehaviors: proto.lapPinBehaviors || [],
    };
  }

  static toProto(config: BartConfig): IBartConfig {
    return {
      name: config.name,
      deviceName: config.deviceName,
      deviceAddress: config.deviceAddress,
      numLanes: config.numLanes,
      minLapMs: config.minLapMs,
      lapPinPitBehavior: config.lapPinPitBehavior,
      lapPinBehaviors: config.lapPinBehaviors,
    };
  }
}

File: server/src/main/java/com/antigravity/converters/BartConfigConverter.java
package com.antigravity.converters;

import com.antigravity.proto.LapPinPitBehavior;
import com.antigravity.protocols.bart.BartConfig;
import java.util.ArrayList;

public class BartConfigConverter {

  public static com.antigravity.proto.BartConfig toProto( // fqn-collision
      BartConfig config) { // fqn-collision
    if (config == null) {
      return com.antigravity.proto.BartConfig.getDefaultInstance(); // fqn-collision
    }
    return com.antigravity.proto.BartConfig.newBuilder() // fqn-collision
        .setName(config.name != null ? config.name : "")
        .setDeviceName(config.deviceName != null ? config.deviceName : "")
        .setDeviceAddress(config.deviceAddress != null ? config.deviceAddress : "")
        .setNumLanes(config.numLanes)
        .setMinLapMs(config.minLapMs)
        .setLapPinPitBehaviorValue(
            config.lapPinPitBehavior != null
                ? config.lapPinPitBehavior.getValue()
                : LapPinPitBehavior.LAP_PIN_PIT_NONE_VALUE)
        .addAllLapPinBehaviors(
            config.lapPinBehaviors != null ? config.lapPinBehaviors : new ArrayList<>())
        .build();
  }

  public static BartConfig fromProto(
      com.antigravity.proto.BartConfig protoConfig) { // fqn-collision
    if (protoConfig == null) {
      return null;
    }
    BartConfig config = new BartConfig();
    config.name = protoConfig.getName();
    config.deviceName = protoConfig.getDeviceName();
    config.deviceAddress = protoConfig.getDeviceAddress();
    config.numLanes = protoConfig.getNumLanes();
    config.minLapMs = protoConfig.getMinLapMs();
    config.lapPinPitBehavior =
        com.antigravity.protocols.arduino.ArduinoConfig.LapPinPitBehavior // fqn-collision
            .fromValue(protoConfig.getLapPinPitBehaviorValue());
    if (protoConfig.getLapPinBehaviorsList() != null) {
      config.lapPinBehaviors = new ArrayList<>(protoConfig.getLapPinBehaviorsList());
    } else {
      config.lapPinBehaviors = new ArrayList<>();
    }
    return config;
  }
}
